src/traces/bar/set_positions.js

```diff
@@ -437,29 +436,12 @@ function updatePositionAxis(gd, pa, sieve, allowMinDtick) {
                 l = p + calcBarOffset,
                 r = l + calcBarWidth;
 
-            if(r >= maxPos) {
-                maxPos = r;
-                pMax = p;
-                pMaxOffset = calcBarOffset;
-                pMaxWidth = calcBarWidth;
-            }
-            if(l <= minPos) {
-                minPos = r;
-                pMin = p;
-                pMinOffset = calcBarOffset;
-                pMinWidth = calcBarWidth;
-            }
+            pMin = Math.min(pMin, l);
+            pMax = Math.max(pMax, r);
         }
     }
 
-    if(pMin) {
-        vpad = pMinWidth / 2;
-        Axes.expand(pa, [pMin + pMinOffset + vpad], {vpad: vpad});
-    }
-    if(pMax) {
-        vpad = pMaxWidth / 2;
-        Axes.expand(pa, [pMax + pMaxOffset + vpad], {vpad: vpad});
-    }
+    Axes.expand(pa, [pMin, pMax], {padded: false});
```

Comment:
Before having `offset` and `width`, the range of the position axis was computed using the position data `x` padded with `vpad = minDiff/2` (`minDiff` being the maximum size of a bar group).

With `offset` and `width`, bars can be shifted away from their positions `x` and widths can be larger than `minDiff`. Here I'm iterating each bar and computing their left and right sides (taking into account `offset` and `width`). This helps me compute the `pMin` and `pMax` needed to expand the position axis, so that all bars are fully within the range of the position axis.

I will add a comment in the code to explain this.